### PR TITLE
change: script bytes helper in toolkit/offchain

### PR DESCRIPTION
**Pull request by AmbientTea**
# Description

Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages.
- [ ] New tests are added if needed and existing tests are updated.
- [ ] Relevant logging and metrics added
- [ ] CI passes. See note on CI.
- [ ] Any changes are noted in the `changelog.md` for affected crate
- [ ] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG Partner Chains developers to do this
for you.

### CODE REVIEW

File: toolkit/offchain/src/plutus_script.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
+use anyhow::{anyhow, Context};
+use cardano_serialization_lib::{Address, LanguageKind, NetworkIdKind, PlutusData};
+use plutus::ToDatum;
+use sidechain_domain::{MainchainAddressHash, PolicyId};
+use uplc::ast::{DeBruijn, Program};
+
+use crate::{csl::*, untyped_plutus::*};
+
+/// Wraps a Plutus script cbor
+pub struct PlutusScript {
```
Comment (LGLO) [inline toolkit/offchain/src/plutus_script.rs:10]: I was not sure about adding our PlutusScript wrapper. I'm not against it. But I think that script should know what version of Plutus it is.

Comment (AmbientTea) [reply toolkit/offchain/src/plutus_script.rs:10]: Added the plutus version. Yeah, I know this is yet-another-wrapper, but we're using two competing plutus data libraries, none of which offers a great API tbh, so I think it's better to hide that ugliness behind a domain-ish object.

File: toolkit/offchain/src/plutus_script.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+			.map_err(|e| anyhow!(e.to_string()))?;
+		program = program.apply_data(data);
+		let bytes = program
+			.to_cbor()
+			.map_err(|_| anyhow!("Couldn't encode resulting script as CBOR."))?;
+		Ok(Self { bytes: bytes.into(), ..self })
+	}
+
+	/// Builds an CSL `Address` for plutus script from the data obtained from smart contracts.
+	pub fn plutus_address(&self, network: NetworkIdKind) -> Address {
```
Comment (LGLO) [inline toolkit/offchain/src/plutus_script.rs:40]: Regarding names of this and following functions, I think that `plutus_` prefix is redundant.

File: toolkit/offchain/src/plutus_script.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+pub struct PlutusScript {
+	pub bytes: Vec<u8>,
+	pub language: LanguageKind,
+}
+
+impl PlutusScript {
+	pub fn from_cbor(cbor: &[u8], language: LanguageKind) -> Self {
+		Self { bytes: cbor.into(), language }
+	}
+	pub fn from_wrapped_cbor(cbor: &[u8], language: LanguageKind) -> anyhow::Result<Self> {
```
Comment (LGLO) [inline toolkit/offchain/src/plutus_script.rs:19]: This one is tricky, for sure rust doc would help this function. I don't know how to stress that it is related to raw-scripts crate.

File: toolkit/offchain/src/plutus_script.rs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+		Ok(Self::from_cbor(&unwrap_one_layer_of_cbor(cbor)?, language))
+	}
+
+	pub fn apply_data(self, data: impl ToDatum) -> Result<Self, anyhow::Error> {
```
Comment (LGLO) [inline toolkit/offchain/src/plutus_script.rs:23]: Are versions in untyped_plutus.rs still required?

Comment (AmbientTea) [reply toolkit/offchain/src/plutus_script.rs:27]: They're not. I removed them and only left casts and unwrapping there.

File: toolkit/offchain/src/plutus_script.rs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	pub bytes: Vec<u8>,
+	pub language: LanguageKind,
+}
+
+impl PlutusScript {
+	pub fn from_cbor(cbor: &[u8], language: LanguageKind) -> Self {
+		Self { bytes: cbor.into(), language }
+	}
+
+	// This function is needed to create [PlutusScript] from scripts in [raw_scripts],
```
Comment (LGLO) [inline toolkit/offchain/src/plutus_script.rs:20]: Cosmetic: if this was `///` comment and you have used backticks, then I think it would be a proper link

**Review by LGLO — APPROVED** [review]

**Pull request merged**